mantidproject/mantid
# Investigate changing the step size for fitting

**Description of work.**
This PR started off by investigating how changing the step size according to the suggestions in #33616 would effect the fit results for various different problems such as muon, ENGINX, crystal field, HIFI and WISH data. I found that the changes had a positive affect on the ENGINX fits, but a slight regression in the fit for the crystal field dataset and a large regression for HIFI data (see the benchmarks in the comments below).

After these results, it was decided it would be useful to be able to switch between using the different step size calculations, so that the user can set the step size calculation method depending on the data they are fitting. With this aim, I have added a property to the `IFittingAlgorithm` class which allows you to change the method used for the step size calculation. There is also some fixes to cppcheck within the associated files.

**To test:**
Make sure the tests are passing.

1. Load the ENGINX calibration data below into mantid
[EnginX.zip](https://github.com/mantidproject/mantid/files/8618788/EnginX.zip)

3. Run this script which uses the usual (default) step size

```
# A python script generated to perform a sequential or simultaneous fit
from mantid.simpleapi import *
import matplotlib.pyplot as plt

# Fit function as a string
function = \
    "name=LinearBackground,A0=0,A1=0;name=BackToBackExponential,I=17.0679,A=1,B=0.05,X0=13642.3,S=1.25655"

# Fitting options
output_base_name = "Output_Default_Fit_"

fit_output = Fit(Function=function, InputWorkspace="EnginX", WorkspaceIndex=0,
                 StartX=13556.2988352, EndX=13731.2988352, MaxIterations=500, StepSizeMethod="Default",
                 Output=output_base_name + "EnginX")

# Plot the results of the fit
fig, axes = plt.subplots(nrows=2,
                         ncols=1,
                         sharex=True,
                         gridspec_kw={"height_ratios": [2, 1]},
                         subplot_kw={"projection": "mantid"})

output_workspace_name = output_base_name + "EnginX_Workspace"
workspace = AnalysisDataService.retrieve(output_workspace_name)
axes[0].plot(workspace, "rs", wkspIndex=0, label="Data", markersize=2)
axes[0].plot(workspace, "b-", wkspIndex=1, label="Fit")
axes[0].set_title(output_workspace_name)
axes[0].set_xlabel("")
axes[0].tick_params(axis="both", direction="in")
axes[0].legend()

axes[1].plot(workspace, "ko", wkspIndex=2, markersize=2)
axes[1].set_ylabel("Difference")
axes[1].tick_params(axis="both", direction="in")

fig.subplots_adjust(hspace=0)
fig.show()
```
After the script finishes, it should open a plot that looks like this:
![image](https://user-images.githubusercontent.com/40830825/166482619-0a409d34-93a6-4669-8860-1d73a2c1557d.png)

3. Run this script which uses the new method for calculating step size:
```
# A python script generated to perform a sequential or simultaneous fit
from mantid.simpleapi import *
import matplotlib.pyplot as plt

# Fit function as a string
function = \
    "name=LinearBackground,A0=0,A1=0;name=BackToBackExponential,I=17.0679,A=1,B=0.05,X0=13642.3,S=1.25655"

# Fitting options
output_base_name = "Output_Sqrt_Fit_"

fit_output = Fit(Function=function, InputWorkspace="EnginX", WorkspaceIndex=0,
                 StartX=13556.2988352, EndX=13731.2988352, MaxIterations=500, StepSizeMethod="Sqrt epsilon",
                 Output=output_base_name + "EnginX")

# Plot the results of the fit
fig, axes = plt.subplots(nrows=2,
                         ncols=1,
                         sharex=True,
                         gridspec_kw={"height_ratios": [2, 1]},
                         subplot_kw={"projection": "mantid"})

output_workspace_name = output_base_name + "EnginX_Workspace"
workspace = AnalysisDataService.retrieve(output_workspace_name)
axes[0].plot(workspace, "rs", wkspIndex=0, label="Data", markersize=2)
axes[0].plot(workspace, "b-", wkspIndex=1, label="Fit")
axes[0].set_title(output_workspace_name)
axes[0].set_xlabel("")
axes[0].tick_params(axis="both", direction="in")
axes[0].legend()

axes[1].plot(workspace, "ko", wkspIndex=2, markersize=2)
axes[1].set_ylabel("Difference")
axes[1].tick_params(axis="both", direction="in")

fig.subplots_adjust(hspace=0)
fig.show()
```

After the script finishes, it should open a plot that looks like this:
![image](https://user-images.githubusercontent.com/40830825/166482910-cb8c6457-5d65-4db2-aa14-11627c6c74b5.png)

Fixes #33616

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
